tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRChecker.scala

```diff
@@ -233,14 +246,63 @@ class IRChecker(unit: LinkingUnit, logger: Logger) {
       }
     }
 
-    if (resultType != AnyType) {
-      reportError(s"Result type of exported method def is $resultType, "+
-          "but must be Any")
+    if (classDef.kind == ClassKind.JSClass && name == "constructor") {
+      checkJSClassConstructor(methodDef, classDef)
+    } else {
+      if (resultType != AnyType) {
+        reportError(s"Result type of exported method def is $resultType, "+
+            "but must be Any")
+      }
+
+      val thisType =
+        if (classDef.kind == ClassKind.JSClass) AnyType
+        else ClassType(classDef.name.name)
+
+      val bodyEnv = Env.fromSignature(thisType, params, resultType)
+      typecheckExpect(body, bodyEnv, resultType)
     }
+  }
 
-    val thisType = ClassType(classDef.name.name)
-    val bodyEnv = Env.fromSignature(thisType, params, resultType)
-    typecheckExpect(body, bodyEnv, resultType)
+  def checkJSClassConstructor(methodDef: MethodDef,
+      classDef: LinkedClass): Unit = {
+    val MethodDef(static, StringLiteral(name), params, resultType, body) = methodDef
```

Comment:
You don't use the name here. Either assert is is `== "constructor"`, or ignore it.